Cross-family review of record — reviewer: kimi (Moonshot, task `review-7112`); author lane: cursor. **FAIL at 948b5ea6db** — one blocking finding (remote shell injection via unquoted `$*` in `_delegate_remote`, reviewer-proven with an ssh shim), three non-blocking residuals. Delta dispatched to the authoring lane (`services-7112-delta`): validate + quote delegated args with a regression test, make the two pid-reconciliation tests xdist-safe, cover the real tunnel-detection helper, document the mixed-tunnel / dropped-flag edges. Re-stamp at the new head, then un-draft and queue.

VERDICT: **FAIL** — PR #7112 @ exact head `948b5ea6dbffbf94d0d687ad7c67e567bbffdf35`

One security finding (remote shell injection into the shared writer host) blocks; everything else checks out.

**1. Syntax / lint**
- `bash -n services.sh` → `SYNTAX OK`. `set -euo pipefail` intact at `services.sh:39`.
- `shellcheck services.sh` (exit 1): two SC2086 (info) at `services.sh:482` and `services.sh:765` (`-iTCP:${SVC_PORT[...]}`), one SC2034 (warning) `FORCE appears unused` at `services.sh:1020`. All three are **pre-existing on `origin/main`** (`git show origin/main:services.sh | shellcheck -` reports the same classes at its lines 413/692/907; FORCE is also write-only there). The PR adds no new shellcheck findings and even annotates its new intentional word-splits with `# shellcheck disable=SC2086`.

**2. Tests**
- Serial: `22 passed, 1 skipped in 19.32s` (skip is the repo-venv-gated live-mode test — expected in a sparse dispatch worktree).
- `-n 2`: `2 failed, 20 passed` — `test_pid_reconciliation` and `test_pid_reconciliation_integration` race on the **real** worktree `.pids/api.pid` (both xdist workers save/delete/restore the same file). Baseline reproduction: I ran `origin/main`'s version of this test file (pre-PR, without the 8 new tests) in-place with `-n 2` → the **same 2 tests fail**. Pre-existing test-isolation flaw (autouse `cleanup_pids_and_logs` fixture + `PIDS_DIR = PROJECT_ROOT/.pids` shared across workers), not introduced by this PR. Noted as residual, not a blocker.

**3. Diff hunt**
- **(b) INJECTION — CONFIRMED, blocking.** Service names from the CLI are never validated before delegation. `_delegate_remote` (`services.sh:327`) interpolates `$*` unquoted into the remote shell string:
  ```
  command ssh -o BatchMode=yes "$host" "cd \"$root\" && ./services.sh ${remote_action}${*:+ $*}"
  ```
  Empirical proof (ssh shim that executes the final arg, `LU_SERVICES_SSH_HOST=fakehost`):
  ```
  $ ./services.sh start 'api;touch /tmp/ssh_inject_test/PWNED'
  Delegating start api;touch /tmp/ssh_inject_test/PWNED to fakehost ...
  sh: line 0: cd: /home/ops/learn-ukrainian: No such file or directory
  $ ls /tmp/ssh_inject_test/
  PWNED  ssh
  ```
  The payload executes on the remote **even though the `cd` failed** (`;` binds looser than `&&`). Pre-PR, a garbage/unknown arg was an inert `Unknown service` no-op; post-PR, whenever delegation triggers (tunnel present or env set — i.e. exactly the target environment), any arg becomes a remote command on shared production. Any wrapper/alias/Makefile forwarding `"$@"` inherits this. Fix is small: validate each requested service against `SVC_CMD` before `_maybe_delegate_and_exit`, and/or build the remote command with `printf '%q'`.
- (a) Misdetection: not found. Detection is per-requested-service (`_requested_has_ssh_tunnel` checks only args, `services.sh:299`), `_is_ssh_tunnel_pid` matches on `comm==ssh` or ssh-leading cmdline (`services.sh:259`), and `_start_service` has the defense-in-depth `_abort_if_ssh_owned_port` guard (`services.sh:638`). Caveat: with no args (`start` = all), one tunneled port delegates *all four* services remotely — documented behavior, but a mixed local/remote setup gets fully remoted. Also detection depends on `lsof`; if lsof is absent, a tunneled port reads as free and local spawn proceeds (the abort guard shares the same blind spot). Residual notes only.
- (c) Remote `stop`: targets only `_ssh_delegate_host` (env or `hramatka` alias), reached only when delegation triggered. Cannot hit an arbitrary host; stopping the writer's services when tunneled is the intended semantics. OK.
- (d) Defaults: `hramatka` applied only inside `_ssh_delegate_host` at ssh time; `LU_SERVICES_REMOTE_ROOT` default assigned at top but consumed only in `_delegate_remote`. `_should_delegate_remote` (`services.sh:313`) triggers solely on explicit env or detected tunnel — confirmed by `test_should_delegate_remote_env_or_tunnel` (`["LOCAL", "ENV", "TUNNEL"]`). OK.
- (e) `status` remains read-only: no delegation hook in the `status` case; it only curls health URLs. OK.

**4. Mutation check**
- Mutation 1 (`return 1` atop `_is_ssh_tunnel_pid`, tunneled port looks local): `test_start_delegates_when_ssh_owns_port` and `test_fix_unhealthy_ssh_tunnel_delegates_restart` **FAILED** (no delegation recorded, local spawn attempted). Note: `test_abort_if_ssh_owned_port_refuses_spawn` still passed here because it self-stubs `_ssh_tunnel_port_pid` — it doesn't cover the real detection helper.
- Mutation 2 (`return 0` atop `_abort_if_ssh_owned_port`, refusal disabled): `test_abort_if_ssh_owned_port_refuses_spawn` **FAILED** (`assert 'REFUSE2' in 'ALLOW\nALLOW2\n'`). Guard is covered.
- Reverted: `git checkout -- .` → `git status --short` empty (`CLEAN`).

**5. Usage/help, dead code, unchanged behavior**
- Header usage block, SSH section, and the `*)` help output all document `fix`, `LU_SERVICES_SSH_HOST`, `LU_SERVICES_REMOTE_ROOT`; `test_usage_documents_fix_and_ssh_env` passes. Verbs match implementation.
- All 7 new helpers are called; no duplicated helpers. `FORCE` unused is pre-existing.
- Ports (8765/8766/8769/4321), health URLs, and Work private-root logic are untouched by the diff — the port-freeze and work-status tests confirm. One silent behavior note: `--live`/`--force` are stripped from `remaining_args`, so a delegated `start api --live` becomes a plain remote `start api` — arguably correct (live mode is a local escape hatch) but undocumented.

**Residual**
1. **Blocker:** unquoted `$*` in `_delegate_remote` → remote shell injection on the writer host (finding 3b). Needs arg validation and/or `printf '%q'` quoting before merge.
2. Pre-existing (not this PR, do not block): `test_pid_reconciliation{,_integration}` are not xdist-safe — they share real `.pids/api.pid`; fails under `-n 2` on `origin/main` too.
3. Coverage gap: the spawn-refusal unit test self-stubs `_ssh_tunnel_port_pid`, so the real `_is_ssh_tunnel_pid` detection path is only covered indirectly via the delegation integration tests.
4. Mixed-tunnel semantics (`start` with only some ports tunneled delegates everything remotely) and lsof-absent blind spot are undocumented edges worth a header note.
5. Delegated calls silently drop `--live`/`--force`.

Worktree is clean, no commits made, no files changed.

--- a/scripts/hygiene/lint_raw_rm_rf.py
+++ b/scripts/hygiene/lint_raw_rm_rf.py
@@ -18,23 +18,26 @@ REPO_ROOT = Path(__file__).resolve().parents[2]
 # Match ``rm -rf``, ``rm -fr``, and compact flag forms such as ``rm -rfv``.
 _RM_RF_RE = re.compile(r"(?:^|[\s;|&])rm\s+-[A-Za-z]*r[A-Za-z]*f[A-Za-z]*\b")
 
-# Exact ``path:line`` allowlist for already-scoped shell cleanups. Do not expand
-# this to cover new recursive deletes — route those through a guarded helper.
-_ALLOWLIST: frozenset[str] = frozenset(
-    {
-        # services.sh: restart lockdir reclaim/release + Astro/Vite cache dirs
-        # (scoped under $PIDS_DIR / $PROJECT_ROOT/site/…). Do not rewrite.
-        "services.sh:144",
-        "services.sh:163",
-        "services.sh:775",
-        "services.sh:785",
-        "services.sh:789",
-        # Scratch-dir EXIT traps for one-shot installer scripts.
-        "scripts/entire/install_kimi_external_agent.sh:18",
-        "scripts/entire/install_fleet_external_agent.sh:36",
-        # Actionlint download cleanup (scoped under a local temp dir).
-        "scripts/audit/check_workflows.sh:53",
-    }
+# Exact ``(path, stripped snippet, max_occurrences)`` allowlist for already-scoped
+# shell cleanups. Matched by content, not line number, so insertions above an
+# entry do not re-break the lint. Do not expand this to cover new recursive
+# deletes — route those through a guarded helper.
+_ALLOWLIST: tuple[tuple[str, str, int], ...] = (
+    # services.sh: restart lockdir reclaim/release + Astro/Vite cache dirs
+    # (scoped under $PIDS_DIR / $PROJECT_ROOT/site/…). Do not rewrite.
+    ("services.sh", "rm -rf \"$lockdir\" 2>/dev/null || true", 2),
+    ("services.sh", "rm -rf \"$vite_cache_dir\"", 1),
+    ("services.sh", "rm -rf \"$dist_dir\"", 1),
+    ("services.sh", "rm -rf \"$astro_dir\"", 1),
+    # Scratch-dir EXIT traps for one-shot installer scripts.
+    ("scripts/entire/install_kimi_external_agent.sh", "trap 'rm -rf \"${scratch_dir}\"' EXIT", 1),
+    ("scripts/entire/install_fleet_external_agent.sh", "trap 'rm -rf \"${scratch_dir}\"' EXIT", 1),
+    # Actionlint download cleanup (scoped under a local temp dir).
+    (
+        "scripts/audit/check_workflows.sh",
+        "cleanup() { [[ -n \"$_DOWNLOAD_DIR\" ]] && rm -rf \"$_DOWNLOAD_DIR\"; return 0; }",
+        1,
+    ),
 )
 
 _SCAN_GLOBS: tuple[str, ...] = (
@@ -54,6 +57,10 @@ def _rel(path: Path, repo_root: Path) -> str:
     return path.resolve().relative_to(repo_root.resolve()).as_posix()
 
 
+def _allowlist_lookup() -> dict[tuple[str, str], int]:
+    return {(path, snippet): max_count for path, snippet, max_count in _ALLOWLIST}
+
+
 def iter_scan_paths(repo_root: Path | None = None) -> list[Path]:
     root = (repo_root or REPO_ROOT).resolve()
     found: list[Path] = []
@@ -70,6 +77,8 @@ def iter_scan_paths(repo_root: Path | None = None) -> list[Path]:
 def find_raw_rm_rf(repo_root: Path | None = None) -> list[tuple[str, str]]:
     """Return ``(path:line, snippet)`` findings not covered by the allowlist."""
     root = (repo_root or REPO_ROOT).resolve()
+    allowlist = _allowlist_lookup()
+    usage: dict[tuple[str, str], int] = {}
     findings: list[tuple[str, str]] = []
     for path in iter_scan_paths(root):
         rel = _rel(path, root)
@@ -80,10 +89,14 @@ def find_raw_rm_rf(repo_root: Path | None = None) -> list[tuple[str, str]]:
         for line_no, line in enumerate(lines, start=1):
             if not _RM_RF_RE.search(line):
                 continue
-            key = f"{rel}:{line_no}"
-            if key in _ALLOWLIST:
-                continue
-            findings.append((key, line.strip()))
+            stripped = line.strip()
+            content_key = (rel, stripped)
+            max_allowed = allowlist.get(content_key)
+            if max_allowed is not None:
+                usage[content_key] = usage.get(content_key, 0) + 1
+                if usage[content_key] <= max_allowed:
+                    continue
+            findings.append((f"{rel}:{line_no}", stripped))
     return findings
 
 

--- a/services.sh
+++ b/services.sh
@@ -25,6 +25,14 @@
 #   start/stop/restart auto-delegate to the writer host when any requested
 #   service port is owned by an ssh listener, or when LU_SERVICES_SSH_HOST
 #   is set. They never spawn or signal local processes in that case.
+#   ``start`` with no service args requests ALL services; when any one of
+#   those ports is tunneled (or LU_SERVICES_SSH_HOST is set), the whole
+#   batch is delegated remotely — not just the tunneled service.
+#   ``--live`` and ``--force`` are local API recovery flags; delegated
+#   start/stop/restart never forwards them to the remote services.sh.
+#   When lsof is absent (or SVC_LSOF_BIN is not executable), port-owner
+#   lookup is a no-op: tunnel auto-delegation and foreign-listener checks
+#   are skipped until a working lsof is available.
 #   fix curls each health URL; unhealthy tunneled ports get a remote
 #   restart. Overrides: LU_SERVICES_SSH_HOST (default host alias hramatka),
 #   LU_SERVICES_REMOTE_ROOT (default /home/ops/learn-ukrainian).
@@ -320,17 +328,28 @@ _should_delegate_remote() {
 _delegate_remote() {
     local remote_action="$1"
     shift
-    local host root
+    local host root remote_cmd arg
     host="$(_ssh_delegate_host)"
     root="${LU_SERVICES_REMOTE_ROOT:-/home/ops/learn-ukrainian}"
+    remote_cmd="./services.sh $(printf '%q' "$remote_action")"
+    for arg in "$@"; do
+        remote_cmd+=" $(printf '%q' "$arg")"
+    done
     echo "Delegating ${remote_action}${*:+ $*} to ${host} (ssh LocalForward or LU_SERVICES_SSH_HOST); not starting or stopping local processes."
-    command ssh -o BatchMode=yes "$host" "cd \"$root\" && ./services.sh ${remote_action}${*:+ $*}"
+    command ssh -o BatchMode=yes "$host" "cd $(printf '%q' "$root") && ${remote_cmd}"
 }
 
 _maybe_delegate_and_exit() {
     local remote_action="$1"
     shift
+    local name
     if _should_delegate_remote "$@"; then
+        for name in "$@"; do
+            if [[ -z "${SVC_CMD[$name]+x}" ]]; then
+                echo "  Unknown service: $name (available: $ALL_SERVICES)"
+                exit 1
+            fi
+        done
         _delegate_remote "$remote_action" "$@"
         exit $?
     fi

--- a/tests/api/test_services_ops.py
+++ b/tests/api/test_services_ops.py
@@ -185,11 +185,24 @@ def cleanup_pids_and_logs():
     elif api_start_file.exists():
         api_start_file.unlink()
 
-def test_pid_reconciliation(temp_services_sh, mock_lsof_env):
+def _patch_script_pids_dir(script_path: Path, pids_dir: Path) -> None:
+    """Point a patched services.sh copy at an isolated pid directory."""
+    content = script_path.read_text(encoding="utf-8")
+    content = content.replace(
+        'PIDS_DIR="$PROJECT_ROOT/.pids"',
+        f'PIDS_DIR="{pids_dir}"',
+    )
+    script_path.write_text(content, encoding="utf-8")
+
+
+def test_pid_reconciliation(temp_services_sh, mock_lsof_env, tmp_path):
     """Test stale pid file / listener interactions hermetically."""
     script_path, port = temp_services_sh
+    pids_dir = tmp_path / "pids"
+    pids_dir.mkdir()
+    _patch_script_pids_dir(script_path, pids_dir)
     set_pids, clear_pids, env = mock_lsof_env
-    api_pid_file = PIDS_DIR / "api.pid"
+    api_pid_file = pids_dir / "api.pid"
 
     # Start a dummy sleep process to act as the listener process.
     proc = subprocess.Popen([
@@ -329,10 +342,13 @@ def test_stop_disables_supervision_before_killing_api_listener(temp_services_sh,
     shutil.which("lsof") is None or sys.platform != "darwin",
     reason="macOS local-ops integration; logic covered hermetically above"
 )
-def test_pid_reconciliation_integration(temp_services_sh_real, mock_lsof_env):
+def test_pid_reconciliation_integration(temp_services_sh_real, mock_lsof_env, tmp_path):
     """Integration test using real sockets and real lsof on macOS."""
     script_path, port = temp_services_sh_real
-    api_pid_file = PIDS_DIR / "api.pid"
+    pids_dir = tmp_path / "pids"
+    pids_dir.mkdir()
+    _patch_script_pids_dir(script_path, pids_dir)
+    api_pid_file = pids_dir / "api.pid"
     set_pids, _, env = mock_lsof_env
 
     # Start a dummy listener process with the API signature configured for our dynamic port
@@ -833,6 +849,64 @@ def _start_named_ssh_process(tmp_path: Path) -> subprocess.Popen[object]:
     return proc
 
 
+def test_is_ssh_tunnel_pid_detects_ssh_and_rejects_non_ssh(tmp_path) -> None:
+    """``_is_ssh_tunnel_pid`` matches ssh comm/argv and rejects other owners."""
+    source = SERVICES_SH.read_text(encoding="utf-8")
+    tunnel = _start_named_ssh_process(tmp_path)
+    non_ssh = subprocess.Popen([str(VENV_PYTHON), "-c", "import time; time.sleep(30)"])
+    reap_process_on_exit(non_ssh)
+    helper = "\n".join(
+        [
+            _extract_bash_function(source, "_cmdline_for_pid"),
+            _extract_bash_function(source, "_is_ssh_tunnel_pid"),
+            f"if _is_ssh_tunnel_pid {tunnel.pid}; then echo SSH_TUNNEL; else echo NOT_SSH; fi",
+            f"if _is_ssh_tunnel_pid {non_ssh.pid}; then echo NON_SSH_MATCH; else echo NON_SSH_OK; fi",
+        ]
+    )
+    try:
+        result = subprocess.run(
+            ["bash", "-c", helper],
+            capture_output=True,
+            text=True,
+            cwd=str(PROJECT_ROOT),
+            timeout=30,
+        )
+        assert result.returncode == 0, result.stderr
+        assert "SSH_TUNNEL" in result.stdout
+        assert "NON_SSH_OK" in result.stdout
+        assert "NON_SSH_MATCH" not in result.stdout
+    finally:
+        tunnel.terminate()
+        non_ssh.terminate()
+        tunnel.wait(timeout=5)
+        non_ssh.wait(timeout=5)
+
+
+def test_delegate_rejects_injection_payload(
+    temp_services_sh, mock_lsof_env, tmp_path
+) -> None:
+    """Shell metacharacters in service args must never reach the ssh delegate."""
+    script_path, _port = temp_services_sh
+    _, _, env = mock_lsof_env
+    shim_dir, capture = _ssh_recorder(tmp_path)
+    env["PATH"] = f"{shim_dir}{os.pathsep}{env.get('PATH', os.defpath)}"
+    env["LU_SERVICES_SSH_HOST"] = "fakehost"
+
+    payload = "api;touch /tmp/ssh_inject_test/PWNED"
+    result = subprocess.run(
+        [str(script_path), "start", payload],
+        capture_output=True,
+        text=True,
+        cwd=str(PROJECT_ROOT),
+        env=env,
+        timeout=30,
+    )
+    combined = f"{result.stdout}\n{result.stderr}"
+    assert "Unknown service" in combined, combined
+    assert result.returncode != 0
+    assert not capture.exists() or capture.read_text(encoding="utf-8") == ""
+
+
 def test_should_delegate_remote_env_or_tunnel() -> None:
     """Delegation triggers on LU_SERVICES_SSH_HOST or an ssh-owned port, not the default alias."""
     source = SERVICES_SH.read_text(encoding="utf-8")
@@ -938,7 +1012,8 @@ def test_start_delegates_when_ssh_host_set(temp_services_sh, mock_lsof_env, tmp_
     args = capture.read_text(encoding="utf-8")
     assert "BatchMode=yes" in args
     assert "hramatka" in args
-    assert 'cd "/home/ops/learn-ukrainian" && ./services.sh start api' in args
+    assert "/home/ops/learn-ukrainian" in args
+    assert "./services.sh start api" in args
     assert "Delegating start api" in result.stdout
     assert _supervisor_was_not_called(env)
 

--- a/tests/test_path_safety.py
+++ b/tests/test_path_safety.py
@@ -131,3 +131,33 @@ def test_raw_rm_rf_lint_detects_unscoped_line(tmp_path: Path) -> None:
 
     findings = lint_raw_rm_rf.find_raw_rm_rf(repo_root=tmp_path)
     assert findings == [("scripts/evil.sh:1", 'rm -rf "$HOME"')]
+
+
+def test_raw_rm_rf_lint_allowlist_keyed_by_content_not_line(tmp_path: Path) -> None:
+    """Allowlisted snippets survive line shifts; excess copies still fail."""
+    lockdir_rm = "rm -rf \"$lockdir\" 2>/dev/null || true"
+    services = tmp_path / "services.sh"
+    services.write_text(
+        "\n".join(
+            [
+                "# preamble inserted above allowlisted cleanups",
+                "# more lines to shift line numbers",
+                lockdir_rm,
+                lockdir_rm,
+                "rm -rf \"$vite_cache_dir\"",
+                "rm -rf \"$dist_dir\"",
+                "rm -rf \"$astro_dir\"",
+            ]
+        )
+        + "\n",
+        encoding="utf-8",
+    )
+
+    findings = lint_raw_rm_rf.find_raw_rm_rf(repo_root=tmp_path)
+    assert findings == [], f"shifted allowlisted snippets should pass: {findings}"
+
+    services.write_text(services.read_text(encoding="utf-8") + lockdir_rm + "\n", encoding="utf-8")
+    findings = lint_raw_rm_rf.find_raw_rm_rf(repo_root=tmp_path)
+    assert len(findings) == 1
+    assert findings[0][1] == lockdir_rm
+    assert findings[0][0].startswith("services.sh:")

